src/realm/util/encrypted_file_mapping.cpp

```diff
@@ -481,14 +486,13 @@ void EncryptedFileMapping::sync() noexcept
 void EncryptedFileMapping::write_barrier(const void* addr, size_t size) noexcept
 {
     REALM_ASSERT(m_access == File::access_ReadWrite);
+    REALM_ASSERT_EX(addr >= m_addr, addr, m_addr);
+    size_t first_accessed_local_page = ((reinterpret_cast<uintptr_t>(addr) - reinterpret_cast<uintptr_t>(m_addr)) >> m_page_shift);
+    size_t num_pages = get_num_pages_for_size(size);
```

Comment:
I think this refactoring is wrong, or at the very least too conservative. To get the right page range, you should still map first and last address separately to page numbers. You try to compute the number of pages solely from the size of the object. I think the address is needed as well to get the "most narrow" result that is still correct.